### DISCUSSION TITLE
Venv-like option to install brew packages within a project (not globally)

### DISCUSSION THREAD

**Discussion by jordiae** [Homebrew/discussions#2950] [Everyday usage]
Hi, I'm looking to do something akin to Python's virtual environments but with Brew:

```
brew venv my_venv
source my_venv/bin/activate
brew install wget  # install wget in my_venv
```

This must be doable because Nix does it, but afaik Brew doesn't support it out of the box. Could someone give me any pointers?

**Comment by gromgit** [comment]
> This must be doable because Nix does it

Nix supports `venv`-like functionality for Homebrew packages? Please post a link to the documentation that states this.

> but afaik Brew doesn't support it out of the box

Yes, and for good reason: Hundreds of Homebrew bottles aren't relocatable, so `brew install` in anything other than the official Homebrew hierarchy quite often results in rebuilding from source, which isn't guaranteed to work on your system, and may take a very long time even if it does.

**Reply by jordiae** [reply]
> > This must be doable because Nix does it
> 
> Nix supports `venv`-like functionality for Homebrew packages? Please post a link to the documentation that states this.

No, Nix supports venv-like functionality for their own packages.

> > but afaik Brew doesn't support it out of the box
> 
> Yes, and for good reason: Hundreds of Homebrew bottles aren't relocatable, so `brew install` in anything other than the official Homebrew hierarchy quite often results in rebuilding from source, which isn't guaranteed to work on your system, and may take a very long time even if it does.

Understood. Still, it's a cost some people would be willing to pay.

**Comment by carlocab** [comment]
A way to do something like this would be to create another installation of Homebrew in a different prefix. See https://github.com/Homebrew/brew/issues/11787.

A quick way to do something like this is
```
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
homebrew/bin/brew install wget
```
Note: this will build non-relocatable bottles (e.g. `openssl@1.1`) from source.

**Reply by jordiae** [reply]
> A way to do something like this would be to create another installation of Homebrew in a different prefix. See [Homebrew/brew#11787](https://github.com/Homebrew/brew/issues/11787).
> 
> A quick way to do something like this is
> 
> ```
> mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
> homebrew/bin/brew install wget
> ```
> 
> Note: this will build non-relocatable bottles (e.g. `openssl@1.1`) from source.

Thanks! I have also found this, what do you think? https://github.com/russtone/brewenv

**Reply by carlocab** [reply]
I haven't encountered it before, but it does look like what you're trying to do.

**Reply by jordiae** [reply]
> I haven't encountered it before, but it does look like what you're trying to do.

Perfect, thank you.
